feat!: robust "firstboot" system

Arcitec opened the pull request:
Everything is deeply tested in a Virtual Machine and works beautifully. The new "firstboot" system is protected (immutable), fixes the issue of legacy leftover scripts, and automatically upgrades all legacy systems to the new link system.

The "justfile" was clashing with "ublue-os/main" in an undesirable way, and it was decided to prefix each justfile with the name of the image/fork, so that has also been fixed here.

People also thought I should include the design flowchart for the new "firstboot" system, so that's also included as a highly compressed little PNG file. Start by [looking at that image](https://github.com/ublue-os/startingpoint/pull/74/commits/44148eb0c21bb6d3996d657436fafad29360333e) and everything will be clear.

This concludes the standardization procedure! After this is merged, we're flexible and fully up to speed with uBlue's direction!

Arcitec committed:
fix!: move all autostart-related files to standardized location

Arcitec committed:
feat: new "autostart" binary, for easy startup customization

Instead of having to use a restrictive command-line in a ".desktop" file, we are now able to use scripting to create more advanced launchers, or if someone wants to easily switch yafti to something else.

Arcitec committed:
fix: switch profile.d script to more robust technique

- Now uses `sh`-style single-bracket `[]` POSIX syntax instead of the truly old-school `test` syntax. Both are equivalent and equally portable. In fact, `man test` explains everything that `[]` can do.

- We now symlink the user's autostart file to the uBlue image's own file, and we do it in a backwards-compatible way which replaces the user's current legacy script file (if any). The previous system dumped a permanent file in the user's home-directory which never updated again after that, which meant that OS updates could never fix or improve the autostart scripts in the future. That's now possible, thanks to being a symlink.

- If the user uninstalls uBlue, or switches to a flavor without `yafti`, then the leftover "autostart" symlink will simply be a minor nuisance that only causes some soft, invisible syslog warnings about the "broken symlink", which is an improvement over the previous system which kept a full `.desktop` file that attempted to run a binary and failed to find it.

Arcitec committed:
feat!: add new recipe flag which allows you to completely disable yafti

This change will allow downstream images to easily opt out of "yafti", if they want to use a different setup program instead.

Arcitec committed:
fix: improve "yq" arguments in justfile

No more pipe redirection. We tell yq to read the file itself. And we use `--` to ensure it never interprets files as flags.

Arcitec committed:
docs: document the new, optional yafti launcher in the README

Arcitec committed:
chore: clean up Containerfile and document "/etc" behavior

- We don't have any files remaining in "./etc/" in this repo.

- We could add a ".gitkeep" file to it, to keep the "empty" directory, but then that file would get automatically merged as a hidden file into the final image, which is silly.

- So let's just document the proper location to place things instead, which is "usr/etc/".

- The Containerfile has also been cleaned up to be a bit easier to follow along what it's doing, such as putting the "copy scripts" step closer to the actual running of the scripts.

Arcitec committed:
feat: include flowchart graphic which explains new launcher hierarchy

castrojo reviewed (approved):
Overall this is awesome! +1 from me but let's give @EinoHR a chance to weigh in since there's a timezone difference that we need to account for. 

Thanks so much for this!

Arcitec committed:
fix!: enhanced autostart.desktop to use new script, and always find icon

Notable changes:

- The icon is no longer dependent on "GNOME Terminal" being installed. Will now work on any desktop environment, not just GNOME. We only see this icon in "GNOME Tweaks: Startup Applications" or similar autostart-managers, which is even better reason to use a generic, always-existing icon for the shortcut.

- Now runs the new "autostart.sh" yafti GUI launcher.

- Tells the OS to not spawn a terminal window when running this shell script.

Arcitec committed:
fix: further improve portability of login-profile.sh

The `-d ""` check is false on bash and zsh, but there might be some rare shell out there which treats it as "." or similar, so let's guard against that risk by also verifying that `$HOME` is a non-empty string before we check if it's a directory.

Arcitec committed:
feat: refactor build.sh to use the new yafti launcher

This implements the new "yafti" launcher and ties everything together:

- The "/etc/profile.d" script is now a symlink to an immutable location, instead of an actual plaintext file, to prevent risk of user systems breaking due to ostree thinking that the script had been modified. We can now be certain that systems always work properly when switching between distros or uBlue flavors.

- Yafti is only installed if the "yafti" flag is enabled.

- If the user disables the "yafti" flag, we completely wipe out the entire yafti.yml configuration, all of our new "yafti autostart" components, and the "profile.d" symlink, so that nothing remains on the system image. If an upstream image installed yafti itself in `/usr/bin`, then we leave that component alone though (because "pip uninstall" and the potential errors it might throw seems like a pointless hassle just for a tiny, unused binary).

Arcitec committed:
fix: friendlier experience in the "yafti" first boot template

- The first screen's "Pick some applications to get started" has been replaced with a friendly welcoming message.

- The second screen's difficult-to-understand "WARNING: This will modify your Flatpaks if you are rebasing!" has been replaced with an explanation of what it actually does.

- The application setup screen is now titled "Application Installer", since the previous title sounded too much like a silly rhyme. It's a minor change.

- All Flatpaks now default to system-wide install thanks to the great work of bsherman at https://github.com/ublue-os/yafti/pull/82. This saves tons of disk space for multi-user systems.

- The "system application" category have been split up into GNOME apps and every other system app, so that people on other desktop environments don't get all the GNOME apps.

- Apps that had too vague descriptions have been renamed to their full names, such as "Backup -> Deja Dup Backups".

- All app lists have been sorted alphabetically.

- Non-inclusive language in descriptions has been changed.

- Added SteamTinkerLaunch as a suggestion for the Steam category, which is the best tool for managing Steam game configurations and Proton installations, albeit very advanced since it can do practically anything the gamer needs. :)

Arcitec commented:
I went through everything one last time with a fine-toothed comb to look for anything that could be improved grammatically or anything else. The last two commits are new, but everything else that "appears to be new" is actually just squashed/slightly reworded files that Jorge has already checked. Phew. It's been a lot of work. Enjoy! :)

EinoHR commented on README.md:
```diff
@@ -21,10 +21,10 @@ The easiest way to start customizing is by looking at and modifying `recipe.yml`
 
 > **Note**
-> The configuration files you put in the `etc` directory are actually added to `/usr/etc/` where they get applied to your local `/etc/` when rebasing to or updating the image. If a config file in `/etc/` has been changed, the changes won't be overridden, but the new version will be available in `/usr/etc/`. Run `sudo ostree admin config-diff` to see the difference between `/etc/` and `/usr/etc/`  ...
+> The configuration files you put in the `usr/etc/` directory are actually added to `/usr/etc/` on the final system, where they will be automatically applied to your local `/etc/` when rebasing to or updating the image. If a config file in `/etc/` has been changed by the user, OSTree will attempt to automatically merge the user's changes into the installed file, but if that fails, the new version ...
```
This note could be slightly rephrased, as it should be obvious that files in `usr/etc/` go to `/usr/etc/`.

Arcitec replied:
@EinoHR Yeah, I followed the flow of the original note but agree that I also felt it was a bit redundant. Fixed. :)

EinoHR commented on README.md:
````diff
@@ -97,9 +97,17 @@ The `latest` tag will automatically point to the latest build. That build will s
+Run the following commands when you're logged into the operating system, to merge uBlue's provided configurations into your own user config. (The "touch" command is only necessary on certain shells which won't let you merge into non-existent files.)
+
+```sh
+touch ~/.justfile
````
Nice!

Arcitec replied:
@EinoHR Yeah, I also force-pushed a change which changes the suffix to `.just` instead, to follow upstream "just" project's standards. We'll soon be able to just include those files beautifully as modules without any of these merging tricks. :)

EinoHR commented on usr/share/ublue-os/firstboot/yafti.yml:
```diff
@@ -67,24 +74,25 @@ screens:
             - Proton Updater for Steam: net.davidotek.pupgui2
         Office:
-          description: Bow to Capitalism
+          description: Boost your productivity.
```
Revert this change NOW :gun:. Oh shoot, that's a water pistol... Well, wouldn't you agree that "Bow to Capitalism" and "Boost your productivity" mean the same thing with a difference in personality...

Arcitec replied:
@EinoHR Yeah it was a bit funny, but we should be neutral and set a good example. :)

EinoHR commented on recipe.yml:
```diff
@@ -1,11 +1,11 @@
-# Warning: only edit this file inside the images git repository.
-# If you are looking at this file in your root directory on your computer, editing it does nothing.
+# Warning: Only edit this file inside the uBlue image's git repository.
```
This warning is useless, as now the recipe resides in an immutable directory.

Arcitec replied:
@EinoHR Very good point, removing the recipe.yml edit warning since it's no longer in a mutable directory. :)

EinoHR commented:
Thanks @Arcitec! I was about to start fixing my complaints, but you don't have *allow maintainers to push commits* on.

Arcitec committed:
fix!: move justfile to standardized, non-conflicting new location

- The old location was conflicting with upstream images (main, nvidia, etc), and was causing the file to be overwritten.

- It was therefore decided that each image should prefix their own justfile names, to avoid clobbering, to easily allow image makers to bundle multiple "modular" justfiles, and to allow end-users to easily include the particular modules they want.

- The name `custom.just` represents a "template name" for this "custom uBlue image", while being neutral enough to use long-term (unlike the alternative name `startingpoint.just`, which doesn't flow nicely).

- All redundant commands that already existed upstream in `ublue-os/main` have been removed, to follow the new "modular inclusion" nature of uBlue's "justfile" organization, which also means that we'll never have to manually update it to match upstream anymore. No more duplicated effort! ;)

- Updated README instructions to mention the new way of including justfiles, until the upstream "just" project has finished their "include" functionality.

- The ".just" suffix is the official upstream suffix for modular justfile inclusions.

Arcitec committed:
docs: clean up leftover comment about editing "recipe.yml"

It's in an immutable location nowadays, so this comment made no sense anymore.

Arcitec commented:
> Thanks @Arcitec! I was about to start fixing my complaints, but you don't have _allow maintainers to push commits_ on.
> 
> Overall the code and changes look fantastic! I'm just ~too~ keen on the documentation side o' things

Hey, great to hear from you! I don't mind the strict feedback at all! You and I both care a lot about clear documentation, and there's no such thing as "caring too much about documentation". :) The users will be happier in the end.

I've addressed the comments, and also did a rebase to fix the `.justfile -> .just` suffix of the modular just file, after discovering that it's how upstream "just" projects officially name things.

So: Two new commits to address the feedback, and an edit of the justfile-renaming commit to use the correct new suffix. That's it! :)

Arcitec committed:
fix: add the missing GNOME Core apps to "yafti" template

- Added missing apps. The list is now synced with upstream's list of Core apps (https://apps.gnome.org/). Minus the few that Fedora doesn't ship by default (such as Console, Music, Web, etc).

- Names that are already installed at the native system-level (such as Nautilus) are not included in the list.

- Fixed the names of a few apps.

- Two of the newly added apps have names that have a high risk of confusing new users, so they were added as "Photos (Organizer)" and "Videos (Player)" to avoid confusion from newcomers, otherwise they might think that those checkboxes add some bundled photos or video files.

Arcitec committed:
fix: remove duplicate flatpaks from recipe.yml

Having duplicates in the yafti.yml where a package exists in multiple groups *currently* causes subtle bugs in yafti, such as a package becoming deselected even though the entire group is "selected". That causes great confusion when users try to install "Core GNOME Apps" without realizing that they won't receive the Extension Manager (for example), since it existed in a different group too and therefore became automatically unchecked in yafti.

A future yafti update will make all duplicate flatpak names illegal (a yml validation error), so this change also prepares us for that, by removing all duplicates.

As for why I added Krita: Since there was only one unique app in the recipe.yml, I needed another random, popular app to fit the example of "this is a selection of example apps", there's nothing more to it than that. ;)

Arcitec commented:
After discovering the bug/quirk in yafti where "duplicate Flatpak names break the installer", I've also cleaned up the duplicate package names now. Spoke to Marco about it on Discord and he'll be marking duplicate yafti yml packages as a validation error, which means that yafti won't even run when there are duplicates in the specification.

There's a ticket here to track that work: https://github.com/ublue-os/yafti/issues/99

Either way, startingpoint's yafti yml is now de-duplicated and will work properly in the future when that change is live.

Arcitec committed:
fix: thoroughly remove fedora Flatpaks, and install unfiltered Flathub

We're now deleting the "fedora" Flatpak repos if they're detected on either the system or user level. All software is crippled in their repos, and less maintained than Flathub, so there's no good reason to have them.

More importantly, we're now also installing the Flathub repo system-wide instead of merely the user-level. This has two important effects:

1. People who opted out of "Third Party Repos" during OS installation will now receive the Flathub repo, to ensure that their system still works properly in that scenario.

2. It gets rid of the "Filter" line in `/var/lib/flatpak/repo/config`, so that Fedora can never reinstate the filtering in the future.

As for why we install Flathub at both the user and system levels: We need both, since the user then gets a nice dropdown menu inside GNOME Software, which lets them switch between system-wide or per-user installation locations.

The "remote" name detection has also been improved, by only printing the raw "name" column, to avoid any risk of false positives.

Arcitec committed:
feat: sanity check Flathub's existence on all systems and all users

This is super important. Without these checks, the system can end up in a broken state where certain users or even the system itself lacks the "flathub" remote, which then breaks yafti's ability to install software, and also breaks consistency between multi-user accounts.

Before these fixes, we only installed "flathub" for the current user on multi-user systems, and then never again, since our installation of Flathub was dependent on the removal of Fedora's own repositories, and never happened again after that.

We now sanity-check both the system and the user-account on every yafti run.

If someone doesn't want "flathub" at the user-level, they're welcome to make their own image which removes that check. But as a "starting point" template, we should strive to provide a robust example that the largest amount of people will be happy with, which means ensuring that Flathub exists for all users on the system!

Most people run a single-user system, and will never even see these screens, since they'll see the initial screen which removes Fedora's repos and takes care of adding Flathub at the same time. These "system repair" screens will only be seen by people on multi-user systems or on various messed up systems.

PS: The strange "run" condition is required since yafti demands that the return code of a command should be "0" to show a conditional screen, but "grep" by itself returns "1" when nothing matches. Negation operators such as `!` don't work in the run-conditions either, so this workaround does the job perfectly instead. It robustly checks for the exact number "0", meaning that the Flathub remote wasn't found.

Arcitec committed:
fix!: remove legacy PWA setup command from custom.just

This is not necessary anymore. It's the job of Flatpak authors. All of them have these permissions by default nowadays, including Microsoft Edge.

Arcitec committed:
docs: clearer explanation of the "/usr/etc" behavior

Including a link to upstream ostree document which explains the config system perfectly.

EinoHR commented:
LGTM.

PS: I'd prefer "why" documentation inline instead of in the commit message, not everyone will be `git blame`-ing their way through the repo to see _why this line is the way it is_.

[see 1eaf70](https://github.com/ublue-os/startingpoint/pull/74/commits/1eaf70e8464aada4929d46189274bb42e5a0aa12)
> PS: The strange "run" condition is required since yafti demands that the return code of a command should be "0" to show a conditional screen, but "grep" by itself returns "1" when nothing matches. Negation operators such as `!` don't work in the run-conditions either, so this workaround does the job perfectly instead. It robustly checks for the exact number "0", meaning that the Flathub remote wasn't found.

EinoHR commented:
Merging either way now.

Arcitec commented:
@EinoHR Yeah, but the comment was in the commit since it was mainly to help you during the review. I was planning to make a ticket @ yafti about the issue instead. Here's the issue tracker entry for it:

https://github.com/ublue-os/yafti/issues/100

If something like that is added, we could clean that line up.

Arcitec commented:
I stumbled on bsherman's repo and saw that he had also gotten rid of the "Bow to Capitalism" line. :smile: 

https://github.com/bsherman/ublue-custom/commit/a07e18a4db848ca9b924a10e3e27cb8d35a29559#diff-79009b18dec43001bbeffb07bdb4b403e07604741cc11578f40400ad25b91628L85

Capitalism wins... this battle. :1st_place_medal:

EinoHR commented:
> Capitalism wins... this battle. 🥇

_noooo_
